tools/codegen/api/python.py

```diff
@@ -210,6 +257,63 @@ def argument_str(self, *, method: bool = False) -> str:
         else:
             return f'{type_str} {name}'
 
+    def argument_str_pyi(self, *, method: bool = False, deprecated: bool = False) -> str:
+        name = self.name
+        if name == 'from':  # from is a Python keyword...
+            name += '_'
+
+        # vestige from 'dynamic_type': remove optional-ness
+        t = self.type
+        if isinstance(t, OptionalType):
+            t = t.elem
+
+        typename = argument_type_str_pyi(t)
+        if self.type.is_nullable():
+            typename = 'Optional[' + typename + ']'
+        if self.default is not None:
+            default = self.default
+            if default == 'nullptr':
+                default = None
+            elif default == 'c10::nullopt':
+                default = None
+            elif default.startswith('{') and default.endswith('}'):
+                if t == BaseType(BaseTy.Tensor) and default == '{}':
+                    default = None
+                elif t == BaseType(BaseTy.Generator) and default == '{}':
+                    default = None
+                elif isinstance(t, ListType) and t.elem == BaseType(BaseTy.int):
+                    default = '(' + default[1:-1] + ')'
+                else:
+                    raise Exception("Unexpected default constructor argument of type {}. default={}, typename={}".format(t, default, type(t)))
+            elif default == 'MemoryFormat::Contiguous':
+                default = 'contiguous_format'
+            elif default == 'QScheme::PER_TENSOR_AFFINE':
+                default = 'per_tensor_affine'
+            default = '={}'.format(default)
+        else:
+            default = ''
+        # TODO: remove after migration (byte-for-byte compatibility)
+        if name == 'requires_grad':
+            colon = ':'
+        else:
+            colon = ': '
+
+        # s/self/input/ outside method bindings
+        # [old codegen] TODO: remove this? doesn't rename in codegen, it's just
+        # for the parse string
+        # this also doesn't apply to deprecated functions for some reason
+        if name == 'self' and typename == 'Tensor' and not method and not deprecated:
+            name = 'input'
+
+        # pyi merges the _out and functional variants into the same signature, with an optional out arg
+        if name == 'out' and typename == 'Tensor' and not deprecated:
+            typename = 'Optional[' + typename + ']'
```

Comment:
I tried digging into whether or not this is really a valid diff between pyi and python. Taking `vdot` as an example:

python arg parser:
```
static PythonArgParser parser({
  "vdot(Tensor input, Tensor other, *, Tensor out=None)",
}, /*traceable=*/true);
```

pyi:
```
def vdot(input: Tensor, other: Tensor, *, out: Optional[Tensor]=None) -> Tensor: ...
```

And the corresponding `vdot` entry in native_functions.yaml has a functional + out variant [here](https://github.com/pytorch/pytorch/blob/93973ee6993c97c8b9d60c4f720423bc625073ea/aten/src/ATen/native/native_functions.yaml#L1550).

It seems like pyi does more of the correct thing than the python arg parser here (typing the `out` argument as optional).

Reply:
Although the signature looks weird, Python arg parser does the "right thing" anyway:

```
>>> torch.vdot(torch.tensor([1]), torch.tensor([2]), out=None)
tensor(2)
```

This is because when an argument in PythonArgParser is defaulted, this automatically flips the "optional" bit:

```
  auto eq = name_str.find('=');
  if (eq != std::string::npos) {
    name = name_str.substr(0, eq);
    optional = true;
    set_default_str(name_str.substr(eq + 1));
  } else { 
    name = name_str;
  }
```

it seems reasonable to change the parsing convention so that optional-ness must be explicitly indicated in the Python arg parser string.